Mathlib/LinearAlgebra/Charpoly/Univ.lean

```diff
@@ -0,0 +1,118 @@
+/-
+Copyright (c) 2024 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+import Mathlib.Data.MvPolynomial.Equiv
+import Mathlib.Data.MvPolynomial.Funext
+import Mathlib.LinearAlgebra.Matrix.Charpoly.Coeff
+import Mathlib.RingTheory.MvPolynomial.Homogeneous
+
+/-!
+# The universal characteristic polynomial
+
+In this file we define the universal characteristic polynomial `Matrix.charpoly.univ`,
+which is the charactistic polynomial of the matrix with entries `Xᵢⱼ`,
+and hence has coefficients that are multivariate polynomials.
+
+It is universal in the sense that one obtains the characteristic polynomial of a matrix `M`
+by evaluating the coefficients of `univ` at the entries of `M`.
+
+We use it to show that the coeffients of the characteristic polynomial
+of a matrix are homogeneous polynomials in the matrix entries.
+
+## Main results
+
+* `Matrix.charpoly.univ`: the universal characteristic polynomial
+* `Matrix.charpoly.univ_map_eval₂Hom`: evaluating `univ` on the entries of a matrix `M`
+  gives the characteristic polynomial of `M`.
+* `Matrix.charpoly.univ_coeff_isHomogeneous`:
+  the `i`-th coefficient of `univ` is a homogeneous polynomial of degree `n - i`.
+-/
+
+open BigOperators
+
+namespace Matrix.charpoly
+
+variable {R S : Type*} (n : Type*) [CommRing R] [CommRing S] [Fintype n] [DecidableEq n]
+variable (f : R →+* S)
+
+variable (R)
+
+/-- The universal characteristic polynomial for `n × n`-matrices,
+is the charactistic polynomial of `Matrix.mvPolynomialX n n ℤ` with entries `Xᵢⱼ`.
+
+Its `i`-th coefficient is a homogeneous polynomial of degree `n - i`,
+see `Matrix.charpoly.univ_coeff_isHomogeneous`.
+
+By evaluating the coefficients at the entries of a matrix `M`,
+one obtains the characteristic polynomial of `M`,
+see `Matrix.charpoly.univ_map_eval₂Hom`. -/
+noncomputable
+abbrev univ : Polynomial (MvPolynomial (n × n) R) :=
+  charpoly <| mvPolynomialX n n R
+
+variable {R}
+
+open MvPolynomial RingHomClass in
+@[simp]
+lemma univ_map_eval₂Hom (M : n × n → S) :
+    (univ R n).map (eval₂Hom f M) = charpoly (Matrix.of M.curry) := by
```

Comment:
Is there a more general result hiding here about `(charpoly A).map f` where `f` is an AlgHom?